### PR TITLE
analyze: several improvements

### LINKED ISSUE #7721: analyze: systemd-analyze doesn't report kernel time
### Submission type

 - Bug report

### systemd version the issue has been seen with

> v230, checked no significant changes of code in v236 

### Used distribution

> custom embedded

### Observation
systemd_7721

`systemd-analyze` doesn't report the kernel time:
`Startup finished in 1.860s (userspace) = 5.957s`

But this is reported in journal by `manager.c`:
`systemd[1]: Startup finished in 4.097s (kernel) + 1.860s (userspace) = 5.957s.`

`gdbus introspect -y -d org.freedesktop.systemd1 -o /org/freedesktop/systemd1 | grep Kernel` returns
```
readonly t KernelTimestamp = 0;
readonly t KernelTimestampMonotonic = 0;
```

At the main.c, directly after `dual_timestamp_from_monotonic(&kernel_timestamp, 0);`, `kernel_timestamp.realtime` is 0

### Expectation
I am not sure if in this case the kernel time shall not be reported by `systemd-analyze`.
Also, the report from `systemd-analyze` and `manager.c` is not consistent. (see above)

### PR DESCRIPTION
**Pull request by yuwata**
This replaces #9790.
Closes #7721.

I've not checked that this works for RTC-less systems. @kmarinushkin Could you check this?

### CODE REVIEW

**Comment by kmarinushkin** [comment]
@yuwata 

It was proved in the PR https://github.com/systemd/systemd/pull/9790 , that the commit 064b52b breaks printing for containers use-case.
I suggest that you remove it from this PR, and I will fix the problem in PR https://github.com/systemd/systemd/pull/9790 and rebase on top of the new HEAD. Most probably this PR will be already merged at that time.

**Comment by yuwata** [comment]
@kmarinushkin The last commit was modified. I've checked that the current version does not break containerized system.

**Comment by kmarinushkin** [comment]
@yuwata oh, nice! than I will test it within the RTC-less system tomorrow and give you a feedback

**Comment by yuwata** [comment]
@kmarinushkin Thanks.

**Comment by yuwata** [comment]
Updated. Now `systemd-analyze -M machine_name plot` also works fine.

File: src/analyze/analyze.c [hunk 1]
```diff
@@ -225,106 +225,47 @@ static void subtract_timestamp(usec_t *a, usec_t b) {
...
+                { "UnitsLoadStartTimestampMonotonic",         "t", NULL, offsetof(struct boot_times, unitsload_start_time)          },
+                { "UnitsLoadFinishTimestampMonotonic",        "t", NULL, offsetof(struct boot_times, unitsload_finish_time)         },
+                { "InitRDSecurityStartTimestampMonotonic",    "t", NULL, offsetof(struct boot_times, initrd_security_start_time)    },
+                { "InitRDSecurityFinishTimestampMonotonic",   "t", NULL, offsetof(struct boot_times, initrd_security_finish_time)   },
+                { "InitRDGeneratorsStartTimestampMonotonic",  "t", NULL, offsetof(struct boot_times, initrd_generators_start_time)  },
+                { "InitRDGeneratorsFinishTimestampMonotonic", "t", NULL, offsetof(struct boot_times, initrd_generators_finish_time) },
+                { "InitRDUnitsLoadStartTimestampMonotonic",   "t", NULL, offsetof(struct boot_times, initrd_unitsload_start_time)   },
+                { "InitRDUnitsLoadFinishTimestampMonotonic",  "t", NULL, offsetof(struct boot_times, initrd_unitsload_finish_time)  },
+        };
```
Comment (keszybz) [inline src/analyze/analyze.c]: This needs `{}` terminator.

File: src/analyze/analyze.c [hunk 2]
```diff
@@ -385,6 +326,12 @@ static void free_host_info(struct host_info *hi) {
 DEFINE_TRIVIAL_CLEANUP_FUNC(struct host_info*, free_host_info);
 
 static int acquire_time_data(sd_bus *bus, struct unit_times **out) {
+        static const struct bus_properties_map property_map[] = {
+                { "InactiveExitTimestampMonotonic",  "t", NULL, offsetof(struct unit_times, activating)   },
+                { "ActiveEnterTimestampMonotonic",   "t", NULL, offsetof(struct unit_times, activated)    },
+                { "ActiveExitTimestampMonotonic",    "t", NULL, offsetof(struct unit_times, deactivating) },
+                { "InactiveEnterTimestampMonotonic", "t", NULL, offsetof(struct unit_times, deactivated)  },
+        };
```
Comment (keszybz) [inline src/analyze/analyze.c]: Here too.

File: src/core/manager.h [hunk 3]
```diff
@@ -53,6 +53,27 @@ typedef enum StatusType {
...
 } StatusType;
 
+/* Notes:
+ * 1. TIMESTAMP_FIRMWARE, TIMESTAMP_LOADER, TIMESTAMP_KERNEL, TIMESTAMP_INITRD,
+ *    TIMESTAMP_SECURITY_START, and TIMESTAMP_SECURITY_FINISH are set only when
+ *    the manager is system and not running under container environment.
+ *
+ * 2. The monotonic timestamp of TIMESTAMP_KERNEL is always zero.
+ *
+ * 3. The realtime timestamp of TIMESTAMP_KERNEL may not set if the system does not
```
Comment (keszybz) [inline src/core/manager.h]: may not set → will be unset

File: src/core/manager.h [hunk 4]
```diff
@@ -53,6 +53,27 @@ typedef enum StatusType {
...
+ *    have RTC.
+ *
+ * 4. TIMESTAMP_FIRMWARE and TIMESTAMP_LOADER may not set if the system does not
```
Comment (keszybz) [inline src/core/manager.h]: may not set → will be unset

File: src/core/manager.h [hunk 5]
```diff
@@ -53,6 +53,27 @@ typedef enum StatusType {
...
+ *    have RTC, or systemd is built without EFI support.
+ *
+ * 5. The monotonic timestamps of TIMESTAMP_FIRMWARE and TIMESTAMP_LOADER are stored as
+ *    negative of the actual value.
+ *
+ * 6. TIMESTAMP_USERSPACE is the timestamp of the manager is started.
```
Comment (keszybz) [inline src/core/manager.h]: Maybe "of when the manager was started" ?

File: src/core/manager.h [hunk 6]
```diff
@@ -53,6 +53,27 @@ typedef enum StatusType {
...
+ *
+ * 7. TIMESTAMP_INITRD_* are set only when the system is booted with initrd.
```
Comment (keszybz) [inline src/core/manager.h]: with *an* initrd

**Comment by kmarinushkin** [comment]
@yuwata 

Looks like we also need this change as a part of the commit 0223938
~~~~
diff --git a/src/analyze/analyze.c b/src/analyze/analyze.c
index 6eb9d39..1becf75 100644
--- a/src/analyze/analyze.c
+++ b/src/analyze/analyze.c
@@ -650,7 +650,7 @@ static int analyze_plot(int argc, char *argv[], void *userdata) {
         }
         if (boot->initrd_time > 0)
                 m++;
-        if (boot->kernel_time > 0)
+        if (boot->kernel_done_time > 0)
                 m++;
 
         for (u = times; u->has_data; u++) {
~~~~

**Comment by yuwata** [comment]
@kmarinushkin OK. Please comment in the review, then I will include them in this PR. Or, it is also welcome to implement that in your original PR.

**Comment by yuwata** [comment]
Ah, I see. I will check what the line do later. (Sorry, we have posted almost the same time. Just ignore my previous comment.)

**Comment by yuwata** [comment]
@kmarinushkin Updated. Now the revised commit includes your suggestion.

**Comment by kmarinushkin** [comment]
@yuwata @keszybz 

I tested this PR WRT printing kernel time on the target board w/o RTC.
From what I observe, this PR fixes the issue https://github.com/systemd/systemd/issues/7721.
Notice that I didn't test the container use-case (for possible missbehaviours, see comment https://github.com/systemd/systemd/pull/9790#issuecomment-410248411).

**Comment by yuwata** [comment]
@kmarinushkin Thank you so much for testing this PR. Let's wait other reviewers' comment.

**Comment by keszybz** [comment]
The `systemd-nspawn` case works properly now.

**Pull request merged**